.github/workflows/cd.yml

```diff
@@ -112,7 +112,57 @@ jobs:
       pull-requests: write
     steps:
       - uses: release-drafter/release-drafter@v5
+        id: update-release-draft
         env:
           GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
         with:
           version: ${{ env.VERSION }}
+
+  build-push-release-binaries:
```

Comment:
Would it be better to cross-build containers and then just take binary out of them? Container image is just a tarball.. This should save half of CI time (we will need both anyways).

Reply:
Hmm, this is probably much faster and less error-prone since it doesn't rely on QEMU, and also lets us build for platforms that Docker doesn't support.

Reply:
If we have caches, it should be pretty fast!

Reply:
@AndrewDryga Ah, I see what you mean. This step happens after we click `Publish release` in the GH UI, so it won't be blocking anything and is Ok to take a few minutes. The uncached run in this PR took about 6 minutes:

https://github.com/firezone/firezone/actions/runs/6538763626/job/17755407225

We could move multi-platform compilation to the `CI` workflow but that would slow things down a lot because I think we'd be building `linux/arm64` on the `x86_64` runners again.

If we stick with this approach we only cross-compile on `publish`.

There could be value in running Integration tests for all platforms but then we'd probably want `arm64` (and whatever other platform) self-hosted Linux runners set up... opened this to track: https://github.com/firezone/firezone/issues/2370

Reply:
That reminds me, I need to add `linux/arm64` Docker image builds in this PR

Reply:
Thinking more about that approach, I don't actually think the `ADD` would be that slow, but I don't think we can (easily) share workflow artifacts across workflows? We'd have to download the binary artifacts from the `ci` workflow and attach it to the drafted release after the `deploy` completes in the `cd` workflow

Reply:
Keeping artifacts used for integration tests and artifacts used for release will be needed anyway, see #2339

Reply:
I think there are a few things we should discuss here separately.

## When should we build production artifacts?

I think it would be better for us to reuse the same containers tested on staging in production, not re-building them on "Publish". Because if we rebuild a container and even re-run acceptance tests it's still a different binary inside, that wasn't ever started on a clone of our production infrastructure (staging) and nobody ever dogfood'ed that release. 

Integration tests, in theory, should catch those, but in practice, we can't cover everything with them. For example, we can have an asset caching issue that might lead to our UI being broken (eg. no JS or CSS).

## Should we run tests on the main?

When you click Publish there should be no failed tests feedback cycle at all, that would be confusing and make pushing hotfixes a much worse experience - imagine something being published on GitHub Releases page but still waiting for 30 minutes to build the artifacts. And then they can fail and we will have to withdraw (delete?) the release, re-create it by re-running `cd.yml`, hotfix whatever failed, and retry.

So I would build everything in `cd.yml` (I believe we can pass arguments from `cd.yml` to `ci.yml` to extend the matrix) and then when you click Publish - we should just pull staging artifacts related to a specific commit this release draft is for, re-label them, and push to GCR, production Artifact Registry, App Store, Google Play, etc.

## How we can optimize cross-compilation?

If Rust can build a binary for a different target platform outside of Docker, then we can build binaries on the Actions runner and then just take an empty [`runtime`](https://github.com/firezone/firezone/blob/main/rust/Dockerfile#L20) image and ADD the binary from the host to a container. This will make us avoid using QEMU, save a ton of time compiling without an additional virtualization layer, and produce both self-contained binaries and containers for everything we need.

Alternatively, we can do it the other way around - building with QEMU in Docker and extracting binary from container image tarball.

This is only needed for Rust stuff, Elixir we should be good having just the amd64 variant as we do now for as long as we don't promote a self-hosting Firezone portal on the arm.

> workflow artifacts across workflows

We already do that between the `main` and PR branches for our migrations: https://github.com/firezone/firezone/blob/main/.github/workflows/elixir.yml#L278.

Reply:
@AndrewDryga Yeah my upcoming changes should address all of this.

The one thing missing with testing images on staging currently is non-amd64 images -- at the moment those will be published with no testing. I think we can leave for later with #2370

Reply:
>  in some other way, eg. only in ci.yml, or on a separate step, not hurting the artifacts reuse?

The issue here is I think compiling with `debug_assertions` creates a different binary and changes the logic, so we inherently need two images. Unless @thomaseizinger -- can `debug_assertions` safely be enabled for `--release`?

> We already do that between the main and PR branches for our migrations:

These are within the same workflow. What we would need to do here is build binary artifacts in `ci.yml`, upload them somewhere, then re-download them in `cd.yml` to attach to the drafted release.

Reply:
> > in some other way, eg. only in ci.yml, or on a separate step, not hurting the artifacts reuse?
> 
> The issue here is I think compiling with `debug_assertions` creates a different binary and changes the logic, so we inherently need two images. Unless @thomaseizinger -- can `debug_assertions` safely be enabled for `--release`?

Yes, they are an entirely different binary. Debug builds have debug symbols in them, are not optimized and have debug-asserts. Release builds are optimised and have no debug assertions.

We can stick with building release images in the containers whether inside or outside isn't that important IMO. Within the container has the advantage that it can easily be reproduced. If we build them outside, I'd go as far as saying we MUST build for musl, otherwise we can run into funky glibc version mismatch issues.

Regarding debug assertions, we could solve this by having more "unit" tests, i.e. Rust tests. We are running those with debug assertions already. The thing with debug assertions is, you don't _want_ to ship them to your customer. They are meant to crash your program on invariant violations that could be fine but are just odd and most likely a bug. It would be really good if we would have an integration test run _somewhere_ that has these enabled. By design, that means these artifacts can't / shouldn't be shipped to customers though.

Can we just run them in parallel maybe and only promote the release binaries to deployment if both the debug and release integration tests pass?

Reply:
@thomaseizinger What we discussed in standup today was the following approach:

- In `ci.yml`, build debug images for automated testing. This allows devs to push up to their PR into a known-good CI env and monitor what fails.
- In `cd.yml`, build release images to test against the `staging` deploy. Currently this is a somewhat manual QA process involving possibly updating Android app, TestFlight build, deploying new Gateways/Relays (potentially on `armv7` and `arm64` machines), etc. Basically anything we can't easily test in the GH actions environment. We can work over time to automate more and more testing here.